osu.Game/Online/SignalRUnionWorkaroundResolver.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using MessagePack;
+using MessagePack.Formatters;
+using MessagePack.Resolvers;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Multiplayer.MatchTypes.TeamVersus;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// Handles SignalR being unable to comprehend [Union] types correctly by redirecting to a known base (union) type.
+    /// See https://github.com/dotnet/aspnetcore/issues/7298.
+    /// </summary>
+    public class SignalRUnionWorkaroundResolver : IFormatterResolver
+    {
+        public static readonly MessagePackSerializerOptions OPTIONS =
+            MessagePackSerializerOptions.Standard.WithResolver(new SignalRUnionWorkaroundResolver());
+
+        private static readonly Dictionary<Type, IMessagePackFormatter> formatter_map = new Dictionary<Type, IMessagePackFormatter>
+        {
+            { typeof(TeamVersusUserState), new TypeRedirectingFormatter<TeamVersusUserState, MatchUserState>() },
+            { typeof(TeamVersusRoomState), new TypeRedirectingFormatter<TeamVersusRoomState, MatchRoomState>() },
+            { typeof(ChangeTeamRequest), new TypeRedirectingFormatter<ChangeTeamRequest, MatchUserRequest>() },
```

Comment:
I'm not really going to bother with looking why the base type spec is needed as it seems like a black hole (and hell knows researching workarounds for this serialisation stuff has taken enough of my time this week already), but have you considered populating this dictionary automatically via assembly scanning/reflection rather than manually? Something like "get all types that have `[Union]` on them, and then register them, as well as the union'd types against the formatter"?

Reply:
We can probably do that as a follow up. Just wanted to get things working for now.